src/key/helper.js

```diff
@@ -4,32 +4,34 @@
  * @private
  */
 
-import {
-  PublicKeyPacket,
-  PublicSubkeyPacket,
-  SecretKeyPacket,
-  SecretSubkeyPacket,
-  SignaturePacket
-} from '../packet';
+import { PublicKeyPacket, PublicSubkeyPacket, SecretKeyPacket, SecretSubkeyPacket, SignaturePacket } from '../packet';
 import enums from '../enums';
 import crypto from '../crypto';
 import util from '../util';
 import defaultConfig from '../config';
 
-export async function generateSecretSubkey(options, config) {
+export async function generateSecretSubkey(options, config, plugin = null) {
+  if (plugin !== null) {
+    plugin.type = 'sub';
+    plugin.dataSubkey = { options, config };
+  }
   const secretSubkeyPacket = new SecretSubkeyPacket(options.date, config);
   secretSubkeyPacket.packets = null;
   secretSubkeyPacket.algorithm = enums.write(enums.publicKey, options.algorithm);
-  await secretSubkeyPacket.generate(options.rsaBits, options.curve);
+  await secretSubkeyPacket.generate(options.rsaBits, options.curve, plugin);
   await secretSubkeyPacket.computeFingerprintAndKeyID();
   return secretSubkeyPacket;
 }
 
-export async function generateSecretKey(options, config) {
+export async function generateSecretKey(options, config, plugin = null) {
+  if (plugin !== null) {
+    plugin.type = 'main';
+    plugin.dataMain = { options, config };
+  }
```

Comment:
I guess this is what you meant by

> Revert serializing the key generation calls (this was needed to make sure that the main key and subkey will not switch places during import).

? Indeed this is quite hacky and should be changed. If you pass the algorithm name, I think that should also be sufficient (instead of passing whether it's the primary key or a subkey), right?

Reply:
That's correct. Instead of this temporary solution I am now passing a new object, which holds the algorithm type as suggested, and restored the concurrent key generation.